decoder/decoder.go

```diff
@@ -0,0 +1,18 @@
+package decoder
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+type ok interface {
+	OK() error
+}
+
+// Decode is a custom decoder validator
+func Decode(r *http.Request, v ok) error {
+	if err := json.NewDecoder(r.Body).Decode(v); err != nil {
+		return err
+	}
+	return v.OK()
+}
```

Comment:
Don't think this is needed - you still have to do `if err != nil` so you aren't saving any duplication.

Reply:
Using this I am able to be sure that there is a required field missing from the JSON.

Reply:
I think thats coupling json deserialisation and validation. If the field is missing that a different error to being unable to unmarshall the JSON.